news-bundle/src/Controller/NewsFeedController.php

```diff
@@ -0,0 +1,103 @@
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\NewsBundle\Controller;
+
+use Contao\CoreBundle\Asset\ContaoContext;
+use Contao\CoreBundle\Controller\AbstractController;
+use Contao\CoreBundle\Routing\Page\DynamicRouteInterface;
+use Contao\CoreBundle\Routing\Page\PageRoute;
+use Contao\CoreBundle\ServiceAnnotation\Page;
+use Contao\NewsBundle\Event\FetchArticlesForFeedEvent;
+use Contao\NewsBundle\Event\TransformArticleForFeedEvent;
+use Contao\PageModel;
+use Contao\StringUtil;
+use FeedIo\Feed;
+use FeedIo\Specification;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+
+/**
+ * @Page(contentComposition=false)
+ */
+class NewsFeedController extends AbstractController implements DynamicRouteInterface
+{
+    public const TYPE = 'news_feed';
+
+    // TODO: Fix content types, maybe make it configurable in the page settings
+    public static $contentTypes = [
+        'atom' => 'application/atom+xml',
+        'json' => 'application/feed+json',
+        'rss' => 'application/rss+xml',
+    ];
+
+    private $urlSuffixes = [
+        'atom' => '.xml',
+        'json' => '.json',
+        'rss' => '.xml',
+    ];
+
+    public function __construct(private readonly ContaoContext $contaoContext, private readonly Specification $specification)
+    {
+    }
+
+    public function __invoke(Request $request, PageModel $pageModel): Response
+    {
+        $this->initializeContaoFramework();
+
+        $staticUrl = $this->contaoContext->getStaticUrl();
+        $baseUrl = $staticUrl ?: $request->getSchemeAndHttpHost();
+
+        $feed = (new Feed())
+            ->setTitle($pageModel->title)
+            ->setDescription($pageModel->description)
+            ->setLanguage($pageModel->language)
+        ;
+
+        $event = new FetchArticlesForFeedEvent($feed, $request, $pageModel);
+        $this->container->get('event_dispatcher')->dispatch($event, FetchArticlesForFeedEvent::NAME);
+
+        if (null === ($articles = $event->getArticles())) {
+            return new Response('', Response::HTTP_NO_CONTENT);
+        }
+
+        foreach ($articles as $article) {
+            $event = new TransformArticleForFeedEvent($article, $feed, $pageModel, $request, $baseUrl);
+            $this->container->get('event_dispatcher')->dispatch($event, TransformArticleForFeedEvent::NAME);
+            $feed->add($event->getItem());
+            $this->tagResponse($article);
+        }
+
+        $formatter = $this->specification->getStandard($pageModel->feedFormat)->getFormatter();
+        $response = new Response($formatter->toString($feed));
+        $response->headers->set('Content-Type', self::$contentTypes[$pageModel->feedFormat]);
+
+        $this->setCacheHeaders($response, $pageModel);
+
+        foreach (StringUtil::deserialize($pageModel->newsArchives, true) as $archive) {
+            $this->tagResponse('contao.db.tl_news_archive.'.$archive);
```

Comment:
it feels a bit strange that this is done by the page controller, but actually fetching the feed items is not. The page controller can't really be sure if these archives are used. Shouldn't this rather be based on the articles that are fetched from the event (their PID)?

Reply:
Yes, you are right. That felt strange all along. I moved it up to the `foreach ($articles)` loop, where I add the tags for the article itself in 360608dca